### PR TITLE
GEOMESA-1140 Use Attr Idx Keys for Query

### PR DESCRIPTION
**Pull request by jahhulbert-ccri**
* Fulfill queries using the keys AND values of the attribute index
  instead of just values

### CODE REVIEW

**Comment by jahhulbert-ccri** [comment]
FYI need to backport to 1.1.x will do once its finalized

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AttributeIdxStrategy.scala [hunk 1]
```diff
@@ -116,19 +125,22 @@ class AttributeIdxStrategy(val filter: QueryFilter) extends Strategy with LazyLo
...
         }
       }
     } else if (descriptor.getIndexCoverage() == IndexCoverage.FULL) {
       // we have a fully encoded value - can satisfy any query against it
-      singleTableScanPlan(sft, filter.secondary, hints.getTransform)
-    } else if (IteratorTrigger.canUseIndexValues(sft, filter.secondary, transform)) {
-      // we can use the index value
+      singleAttrValueOnlyPlan(sft, filter.secondary, hints.getTransform)
+    } else if (IteratorTrigger.canUseAttrIdxValues(sft, filter.secondary, transform)) {
+      // we can use the index value. Transform can only include
```
Comment (elahrvivaz) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/AttributeIdxStrategy.scala]: partial comment?

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/AttrKeyPlusValueIterator.scala [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    // Setter functions to pull values from the index SF into the projected trans SF
+    import scala.collection.JavaConversions._
+    val setters: Seq[(KryoBufferSimpleFeature, ScalaSimpleFeature) => Unit] =
+      transSft.getAttributeDescriptors
+        .map(_.getLocalName)
+        .filter(n => idxSft.getAttributeDescriptors.exists(_.getLocalName == n))
+        .map { name =>
+          val transIdx = transSft.getAttributeDescriptors.indexWhere(_.getLocalName == name)
+          val indexIdx = idxSft.getAttributeDescriptors.indexWhere(_.getLocalName == name)
```
Comment (elahrvivaz) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/AttrKeyPlusValueIterator.scala]: [minor] you can just do idxSft.indexOf(name)

**Comment by elahrvivaz** [comment]
looks good - do we care about bin queries or stat aggregating queries?

**Comment by elahrvivaz** [comment]
[idly] i wonder if it would be worthwhile to use the regular kryo iterator, and stack another iterator on top that adds the attribute value to the return feature. would be slower (has to transform twice), but would keep the codebase simpler, and might be ok for the relatively small-ish queries this use case is targetting.

**Comment by jahhulbert-ccri** [comment]
@elahrvivaz yeah i was wondering the same thing about stacking itrs. I wonder how bad the perf hit would be. Expected result sizes range generally from 50 to 800 to upwards of 50k entries returned. Usually most are are expected to be 1k-10k. If anyone else reads this maybe they have an opinion. Either is OK for me. I'll look at what it takes to stack...It would definitely make it simpler.

We do care about bin and stat aggregating queries in the long term for sure. I'll prob make a ticket for that though. We can always refactor that in.

**Comment by elahrvivaz** [comment]
b9afce00075274441cae7744e0cbe1816cd6b0d9

**Pull request closed without merging**